File: code_puppy/plugins/agent_trace/store.py
"""Agent Trace V2 — NDJSON Persistence.

Simple append-only storage for trace events. Events are written
as newline-delimited JSON for easy streaming and replay.
"""

from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Iterator

from code_puppy.config_paths import (
    assert_write_allowed,
    resolve_path,
    safe_append,
    safe_mkdir_p,
    safe_rm,
)
from code_puppy.plugins.agent_trace.schema import TraceEvent

logger = logging.getLogger(__name__)


class TraceStore:
    """Append-only NDJSON store for trace events.

    Events are stored in files named by trace_id:
        ~/.code_puppy/traces/{trace_id}.ndjson

    This enables:
    - Replay of past runs
    - Comparison between runs
    - Post-hoc analysis and aggregation
    """

    def __init__(self, base_dir: Path | str | None = None):
        """Initialize the trace store.

        Args:
            base_dir: Directory for trace files. Defaults to ~/.code_puppy/traces/
        """
        if base_dir is None:
            base_dir = resolve_path("traces")

        assert_write_allowed(base_dir, "trace_store_init")
        self.base_dir = Path(base_dir)
        self._ensure_dir()

    def _ensure_dir(self) -> None:
        """Ensure the trace directory exists."""
        try:
            safe_mkdir_p(self.base_dir)
        except OSError as e:
            logger.warning(f"Failed to create trace directory: {e}")

    def _trace_path(self, trace_id: str) -> Path:
        """Get the file path for a trace."""
        # Sanitize trace_id for filesystem safety
        safe_id = "".join(c if c.isalnum() or c in "-_" else "_" for c in trace_id)
        return self.base_dir / f"{safe_id}.ndjson"

    def append(self, event: TraceEvent) -> bool:
        """Append an event to its trace file.

        Returns True if successful, False otherwise.
        """
        if not event.trace_id:
            logger.debug("Cannot store event without trace_id")
            return False

        try:
            path = self._trace_path(event.trace_id)
            safe_append(path, event.to_json() + "\n", encoding="utf-8")
            return True
        except OSError as e:
            logger.debug(f"Failed to append event: {e}")
            return False

    def append_batch(self, events: list[TraceEvent]) -> int:
        """Append multiple events, grouped by trace_id.

        Returns count of successfully written events.
        """
        # Group by trace_id for efficient file access
        by_trace: dict[str, list[TraceEvent]] = {}
        for event in events:
            if event.trace_id:
                by_trace.setdefault(event.trace_id, []).append(event)

        written = 0
        for trace_id, trace_events in by_trace.items():
            try:
                path = self._trace_path(trace_id)
                safe_append(
                    path,
                    "".join(event.to_json() + "\n" for event in trace_events),
                    encoding="utf-8",
                )
                written += len(trace_events)
            except OSError as e:
                logger.debug(f"Failed to write batch for {trace_id}: {e}")

        return written

    def read(self, trace_id: str) -> list[TraceEvent]:
        """Read all events for a trace.

        Returns empty list if trace doesn't exist or can't be read.
        """
        path = self._trace_path(trace_id)
        if not path.exists():
            return []

        events = []
        try:
            with open(path, "r", encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if line:
                        try:
                            events.append(TraceEvent.from_json(line))
                        except (json.JSONDecodeError, KeyError) as e:
                            logger.debug(f"Skipping malformed event: {e}")
        except OSError as e:
            logger.debug(f"Failed to read trace {trace_id}: {e}")

        return events

    def stream(self, trace_id: str) -> Iterator[TraceEvent]:
        """Stream events for a trace (memory-efficient for large traces)."""
        path = self._trace_path(trace_id)
        if not path.exists():
            return

        try:
            with open(path, "r", encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if line:
                        try:
                            yield TraceEvent.from_json(line)
                        except json.JSONDecodeError, KeyError:
                            continue
        except OSError:
            return

    def list_traces(self) -> list[str]:
        """List all available trace IDs."""
        try:
            return [p.stem for p in self.base_dir.glob("*.ndjson")]
        except OSError:
            return []

    def delete(self, trace_id: str) -> bool:
        """Delete a trace file."""
        path = self._trace_path(trace_id)
        try:
            if path.exists():
                safe_rm(path)
            return True
        except OSError as e:
            logger.debug(f"Failed to delete trace {trace_id}: {e}")
            return False

    def size(self, trace_id: str) -> int:
        """Get the size of a trace file in bytes."""
        path = self._trace_path(trace_id)
        try:
            return path.stat().st_size if path.exists() else 0
        except OSError:
            return 0

    def event_count(self, trace_id: str) -> int:
        """Count events in a trace (requires reading the file)."""
        path = self._trace_path(trace_id)
        if not path.exists():
            return 0

        try:
            with open(path, "r", encoding="utf-8") as f:
                return sum(1 for line in f if line.strip())
        except OSError:
            return 0

File: lib/code_puppy_control/plugins/agent_trace/store.ex
defmodule CodePuppyControl.Plugins.AgentTrace.Store do
  @moduledoc """
  NDJSON persistence for trace events.

  Append-only storage with files named by trace_id under
  `~/.code_puppy_ex/traces/`. All writes go through
  `Config.Isolation.safe_write!`.
  """

  alias CodePuppyControl.Config.Paths
  alias CodePuppyControl.Plugins.AgentTrace.Schema

  @doc "Get the base directory for trace files."
  @spec base_dir() :: String.t()
  def base_dir, do: Path.join(Paths.data_dir(), "traces")

  @doc "Append an event to its trace file."
  @spec append(Schema.trace_event()) :: boolean()
  def append(event) do
    if event.trace_id == "" do
      false
    else
      path = trace_path(event.trace_id)
      dir = Path.dirname(path)
      File.mkdir_p!(dir)
      line = Schema.to_json(event) <> "\n"

      case File.write(path, line, [:append]) do
        :ok -> true
        {:error, _} -> false
      end
    end
  end

  @doc "Append multiple events at once."
  @spec append_batch([Schema.trace_event()]) :: non_neg_integer()
  def append_batch(events) do
    events
    |> Enum.group_by(& &1.trace_id)
    |> Enum.reduce(0, fn {trace_id, trace_events}, acc ->
      path = trace_path(trace_id)
      dir = Path.dirname(path)
      File.mkdir_p!(dir)
      content = Enum.map_join(trace_events, "\n", &Schema.to_json/1) <> "\n"

      case File.write(path, content, [:append]) do
        :ok -> acc + length(trace_events)
        {:error, _} -> acc
      end
    end)
  end

  @doc "Read all events for a trace."
  @spec read(String.t()) :: [Schema.trace_event()]
  def read(trace_id) do
    path = trace_path(trace_id)

    case File.read(path) do
      {:ok, content} ->
        content
        |> String.split("\n", trim: true)
        |> Enum.map(fn line ->
          try do
            Schema.from_json(line)
          rescue
            _ -> nil
          end
        end)
        |> Enum.reject(&is_nil/1)

      {:error, _} ->
        []
    end
  end

  @doc "List all available trace IDs."
  @spec list_traces() :: [String.t()]
  def list_traces do
    dir = base_dir()

    if File.dir?(dir) do
      dir
      |> File.ls!()
      |> Enum.filter(&String.ends_with?(&1, ".ndjson"))
      |> Enum.map(&String.replace_suffix(&1, ".ndjson", ""))
    else
      []
    end
  end

  @doc "Delete a trace file."
  @spec delete(String.t()) :: boolean()
  def delete(trace_id) do
    path = trace_path(trace_id)

    case File.rm(path) do
      :ok -> true
      {:error, _} -> false
    end
  end

  @doc "Count events in a trace."
  @spec event_count(String.t()) :: non_neg_integer()
  def event_count(trace_id) do
    path = trace_path(trace_id)

    case File.read(path) do
      {:ok, content} ->
        content |> String.split("\n", trim: true) |> length()

      {:error, _} ->
        0
    end
  end

  defp trace_path(trace_id) do
    safe_id = trace_id |> String.replace(~r/[^a-zA-Z0-9._-]/, "_")
    Path.join(base_dir(), "#{safe_id}.ndjson")
  end
end
